fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/mixin/networking/ClientConnectionMixin.java

```diff
@@ -63,16 +63,27 @@ private void initAddedFields(NetworkSide side, CallbackInfo ci) {
 		this.playChannels = Collections.newSetFromMap(new ConcurrentHashMap<>());
 	}
 
+	@Unique
+	private Throwable caughtException;
+
+	// Must be fully qualified due to mixin not working in production without it
+	@SuppressWarnings("UnnecessaryQualifiedMemberReference")
+	@Inject(method = "Lnet/minecraft/network/ClientConnection;exceptionCaught(Lio/netty/channel/ChannelHandlerContext;Ljava/lang/Throwable;)V", at = @At(value = "INVOKE", target = "Lnet/minecraft/network/ClientConnection;send(Lnet/minecraft/network/Packet;Lio/netty/util/concurrent/GenericFutureListener;)V"))
+	private void resendOnExceptionCaught$captureThrowable(ChannelHandlerContext context, Throwable ex, CallbackInfo ci) {
+		caughtException = ex;
+	}
+
 	// Must be fully qualified due to mixin not working in production without it
 	@SuppressWarnings("UnnecessaryQualifiedMemberReference")
 	@Redirect(method = "Lnet/minecraft/network/ClientConnection;exceptionCaught(Lio/netty/channel/ChannelHandlerContext;Ljava/lang/Throwable;)V", at = @At(value = "INVOKE", target = "Lnet/minecraft/network/ClientConnection;send(Lnet/minecraft/network/Packet;Lio/netty/util/concurrent/GenericFutureListener;)V"))
 	private void resendOnExceptionCaught(ClientConnection self, Packet<?> packet, GenericFutureListener<? extends Future<? super Void>> listener) {
 		PacketListener handler = this.packetListener;
+		Text disconnectMessage = new TranslatableText("disconnect.genericReason", "Internal Exception: " + caughtException);
```

Comment:
```suggestion
	private void resendOnExceptionCaught(ClientConnection self, Packet<?> packet, GenericFutureListener<? extends Future<? super Void>> listener, ChannelHandlerContext context, Throwable ex) {
		PacketListener handler = this.packetListener;
		Text disconnectMessage = new TranslatableText("disconnect.genericReason", "Internal Exception: " + ex);
```

The above inject is unnecessary as [`Redirect`](https://jenkins.liteloader.com/job/Mixin/javadoc/org/spongepowered/asm/mixin/injection/Redirect.html) can also capture the target method arguments.

Reply:
Might want to use `ex.getMessage()` as well. Is it really safe to leak the stack trace to the connecting clients?

Reply:
Agreed, it would likely be better to log the full trace, but only send the message to the client.

Reply:
No ex is fine (involves ex.toString which is just the class name + message) - some exceptions have to message and thus would just show as `null`, which is pretty bad.

Reply:
> The above inject is unnecessary as [`Redirect`](https://jenkins.liteloader.com/job/Mixin/javadoc/org/spongepowered/asm/mixin/injection/Redirect.html) can also capture the target method arguments.

Oh, I didn't realize.